### DISCUSSION TITLE
KNI Web - Fixed resolution?

### DISCUSSION THREAD

**Discussion by trinith** [MonoGame/MonoGame#8391] [General]
I'm not sure the best place to put this so I'll drop it here and tag @nkast and see how it goes!

I'm fairly familiar with MonoGame but not at all familiar with web development. However, I'm experimenting with a few things and would like the option of hosting on itch.io. A bit of googling showed me that NKast's KNI Engine is a good option for a web deployment. Unfortunately I'm having a bit of trouble understanding how screen resolution works in this context.

I typically like making 2D retro games so I'll render to a RenderTarget2D and scale up, at fixed integer multiples, to my device's resolution. On desktop I would usually just set this resolution using...

```
_graphics.PreferredBackBufferWidth = 1280;
_graphics.PreferredBackBufferHeight = 720;
_graphics.ApplyChanges();
```

When I do this in a web game I'm getting strange results in that the world space I expect to be rendering to is shifted down maybe 20 pixels. If I resize the browser window the game goes back setting the resolution to whatever the browser's control area is (or whatever the web equivalent for that is).

I'm pretty sure I'm going about this all wrong, but what I would like is a fixed resolution within my browser window. Using the example above, how would I ensure a resolution of 1280x720 in my web game? Are there any suggestions/guidance here?

Thanks for any help with this :)

**Comment by trinith** [comment]
I think I figured this out. Instead of setting the resolution, if I use viewport instead, then I get the desired behaviour. The game will still render to the browser's canvas resolution but I can fix the resolution my own drawing renders to so that everything looks correct. If the browser window is too small things are just cut off, which is fine with me. I have this in `Initialize`...

```
protected override void Initialize()
{
    //_graphics.PreferredBackBufferWidth = 1280;
    //_graphics.PreferredBackBufferHeight = 720;
    //_graphics.ApplyChanges();

    _graphics.GraphicsDevice.Viewport = new Viewport(0, 0, 1280, 720);
    this.Window.ClientSizeChanged +=
        (sender, e) =>
        {
            _graphics.GraphicsDevice.Viewport = new Viewport(0, 0, 1280, 720);
        };

    base.Initialize();
}
```

If there's a better way to go about this though, I'd certainly appreciate some feedback :)

**Reply by nkast** [reply]
You can't resize the browser window.
https://stackoverflow.com/questions/7022787/can-i-resize-the-browser-window

If you need to have the game run in a fixed size, you can host the game inside an iframe.
...or use Resolution Independent Rendering.

**Reply by trinith** [reply]
I'm not sure what an iframe is or how it works, but I can do some googling. The code I posted above will work for me and will help me get to resolution indepdent rendering because I'll still want to maintain pixel squareness. I think I'm good for now but an iframe is probably what I'm after if I publish on itch.io.

Thanks :)

**Reply by nkast** [reply]
itch.io will load the game inside an iframe for you.
All you have to do is to specify the resolution in the upload page.

**Comment by trinith** [comment]
@mrhelmut I'm not sure who to contact about this but the power above copy/pasted my post and injected a few links that I can only assume are phising. I don't see a "report this post" option like we used to have on the old community site. What's the appropriate course of action surrounding stuff like this?
